lib/spack/spack/directives.py

```diff
@@ -189,15 +196,44 @@ def _decorator(decorated_function):
 
             @functools.wraps(decorated_function)
             def _wrapper(*args, **kwargs):
+                # If any of the arguments are executors returned by a
+                # directive passed as an argument, don't execute them
+                # lazily. Instead, let the called directive handle them.
+                # This allows nested directive calls in packages.  The
+                # caller can return the directive if it should be queued.
+                def remove_directives(arg):
+                    directives = DirectiveMetaMixin._directives_to_be_executed
+                    if isinstance(arg, (list, tuple)):
```

Comment:
Not sure I understand what you want here -- `args` is outside the function scope and if an arg happens to be a collection, we still need to descend into it...

Reply:
Also are there other examples of directives as arguments besides passing `patch` to `depends_on`? Either way IMO it would make sense to describe that particular case in the explanatory comment.